adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -891,42 +892,68 @@ protected void onCommandReceived(final TenantObject tenantObject, final ProtonSe
             MessageHelper.addDeviceId(msg, command.getOriginalDeviceId());
         }
 
-        // TODO time out waiting for disposition update
+        final AtomicBoolean isCommandSettled = new AtomicBoolean(false);
+
+        final Long timerId = getConfig().getSendMessageToDeviceTimeout() < 1 ? null :
+            vertx.setTimer(getConfig().getSendMessageToDeviceTimeout(), tid -> {
+            log.debug("waiting for delivery update timed out after " + getConfig().getSendMessageToDeviceTimeout() + " ms");
+            if (isCommandSettled.compareAndSet(false, true)) {
+                final Exception ex = new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE,
+                        "timeout waiting for delivery update");
+                closeLinkWithError(sender, ex, commandContext.getCurrentSpan());
+                // timeout expired -> release command
+                commandContext.getCurrentSpan().log("timeout waiting for delivery update from device");
+                commandContext.release();
+            } else {
+                log.trace("command is already settled and downstream application notified");
+            }
+        });
```

Comment:
Can you run the code-formatter here? I think the code will be better readable then.